### PR TITLE
[python][Feat] Deserialize error responses

### LINKED ISSUE #16967: [BUG][Python] Pydantic 2 generated client does not properly handle "204 No Content"  if it is the first response in a path -- empty `_response_types_map`
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?

##### Description

When using a valid "204 No Content" response as the _first_ response to a path, none of the responses for that path will be included in the `_response_types_map` for that endpoint.

Please see the provided spec for a reproducible example.

If I have an endpoint `/will-fail` that returns (in this order):
- 204 No Content
- 401 Some Content
- 404 Some Content

None of the responses will be correctly handled by the generated python client.

One issue this causes:
When I get a 401, an `open_api.exceptions.Unauthorized` exception is thrown instead of getting my spec's 401 response.

##### openapi-generator version

latest as of 2023-11-01 and 2023-11-02. This also happens in v7.0.0 and v7.0.1

##### OpenAPI declaration file content or url

```yaml
openapi: 3.0.0
info:
  version: 1.0.0
  title: Sample API
  description: A sample API to illustrate 204 No Content issue
paths:
  /will-not-fail:
    delete:
      description: Returns nothing
      responses:
        "200":
          description: Successful response
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/SuccessMessage"
        "404":
          description: Not found
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/NotFound"
  /will-fail:
    delete:
      description: Returns nothing
      responses:
        "204":
          description: Successful response
        "404":
          description: Not found
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/NotFound"
  /will-strangely-not-fail:
    delete:
      description: Returns nothing
      responses:
        "200":
          description: Successful response
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/SuccessMessage"
        "204":
          description: Successful response
        "404":
          description: Not found
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/NotFound"
components:
  schemas:
    NotFound:
      type: object
      properties:
        message:
          type: string
          example: Not found
    SuccessMessage:
      type: object
      properties:
        message:
          type: string
          example: Success
```

To generate, I'm using something like the following:

```sh
rm -rf out/python

docker run --rm -v "${PWD}:/local" openapitools/openapi-generator-cli generate \
    -i "/local/sample-spec.yaml" \
    -g python \
    -o /local/out/python
```

##### Steps to reproduce

If you run the generator on the spec and look at `default_api.py`, you will see that for the `will-fail` path, it has no responses in `_response_types_map`, which causes problems. If you look at the `will-strangely-not-fail` and `will-not-fail` paths, there are the proper responses in `_response_types_map` even though `will-strangely-not-fail` also has a 204 No Content response.

##### Suggest a fix

I will see if I can fix this myself, but I cannot guarantee it. Thank you all for all of your hard work.

### PR DESCRIPTION
**Pull request by robertschweizer**
This adds deserialized error responses to the raised HTTP exceptions, as I suggested in https://github.com/OpenAPITools/openapi-generator/pull/16999

Closes #16967

See separate commits.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@wing328 @fa0311 @multani

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/python/partial_api.mustache [hunk 1]
```diff
@@ -44,11 +44,9 @@
         )
 
         _response_types_map: Dict[str, Optional[str]] = {
-            {{#returnType}}
             {{#responses}}
             {{^isWildcard}}
             '{{code}}': {{#dataType}}"{{.}}"{{/dataType}}{{^dataType}}None{{/dataType}},
             {{/isWildcard}}
             {{/responses}}
-            {{/returnType}}
```
Comment (robertschweizer) [inline modules/openapi-generator/src/main/resources/python/partial_api.mustache]: What was the purpose of the `{{#returnType}}` condition here?

**Comment by robertschweizer** [comment]
@multani @wing328 @fa0311 do any of you have time for a review?

File: modules/openapi-generator/src/main/resources/python/partial_api.mustache (hunk 1 above)
Comment (fa0311) [reply modules/openapi-generator/src/main/resources/python/partial_api.mustache]: This does not seem to affect the operation.
The `{{#returnType}}` simplifies the generated source code, but I don't think the simplification is necessary.

**Review by fa0311 — APPROVED** [review]
Looks good to me.

File: modules/openapi-generator/src/main/resources/python/partial_api.mustache (hunk 1 above)
Comment (robertschweizer) [reply modules/openapi-generator/src/main/resources/python/partial_api.mustache]: Awesome, thanks for checking!

**Pull request merged**